lib/exceptional_synchrony/callback_exceptions.rb

```diff
@@ -38,6 +38,40 @@ def return_exception(*args)
           ex
         end
       end
+
+      private
+
+      def handle_result_error(result)
+        error = result.error
+        if error_is_a_timeout?(error)
+          raise Timeout::Error
+        elsif error_is_a_syscall_class?(error)
+          raise Failure, format_failure(error.to_s, format_inspect(result))
+        else
+          raise Failure, format_failure(format_inspect(error), format_inspect(result))
+        end
+      end
+
+      def format_failure(error_string, result_string)
+        "ERROR = #{error_string}; RESULT = #{result_string}"
+      end
+
+      def format_inspect(obj)
+        inspection = obj.inspect[0..100]
+        if inspection.length > 100
+          inspection[0..91] + '...TRUNC'
+        else
+          inspection
+        end
+      end
+
+      def error_is_a_timeout?(error)
+        error =~ /timeout/i || error == Errno::ETIMEDOUT
+      end
+
+      def error_is_a_syscall_class?(error)
+        error.is_a?(Class) && error.superclass == SystemCallError
```

Comment:
The Liskov Substitution Principle says that sub-sub-classes should work too. I like `<=` for this: It builds on the `<` notation for inheritance; it's like `is_a?` at the class level:
```
error.is_a?(Class) && error <= SystemCallError
```